Metada on session start action

lluchini opened the pull request:
**Proposed changes**:
Closes #5574

- The default action 'action_session_start' wasn't able to receive the metadata from the message sent, RASA when initiating a new session it sends a blank tracker with no information whatsoever, ignoring the metadata into the message.

Adjusted to add the event sessionStart to tracker events and with it the metadata from the message.

**Status (please check what you already did)**:
- [X] added some tests for the functionality
- [X] updated the documentation
- [X] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [X] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RasaHQ/rasa?pullRequest=5611) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/RasaHQ/rasa?pullRequest=5611) before we can accept your contribution.<br/><hr/>**lluchini** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RasaHQ/rasa?pullRequest=5611) it.</sub>

sara-tagger commented:
Thanks for submitting a pull request 🚀 @degiz will take a look at it as soon as possible ✨

degiz commented on tests/core/test_actions.py:
```diff
@@ -510,6 +510,21 @@ async def test_action_session_start_without_slots(
     assert events == [SessionStarted(), ActionExecuted(ACTION_LISTEN_NAME)]
 
 
+async def test_action_session_start_with_metada(
```
```suggestion
async def test_action_session_start_with_metadata(
```

degiz replied:
Also, I think a unit test for this change should be more "high level", since we've modified the code all the way from `log_message()`

degiz commented on rasa/core/processor.py:
```diff
@@ -193,6 +198,9 @@ async def get_tracker_with_session_start(
         if not tracker:
             return None
 
+        if metadata:
```
I think it makes sense to put that into `_update_tracker_session `, after `if` statement on the line 169. Then it's very clear that we add metadata only for the first message that triggers the session.
What do you think?

lluchini replied:
You're right!

This is my first PR on an open-source project, and this is my first implementation in Python, had a lot on my head so I missed this completely, sorry.

lluchini commented:
> > I didn't get what you meant by more "high level".
> 
> The most high level (in terms of structure) code we've modified in this PR is `get_tracker_with_session_start()`. So I'd expect the unit tests that cover this change to call this method with `metadata` and assert the results.

I hope that I had understand what you meant, I added a test into teste_processor.py to this specific method and assertion of the metadata content.

lluchini commented:
The deep source is failing with a false positive.

lluchini commented:
**Faling Not My Fault :)**

> Continuous Integration / Build Docker (docker/Dockerfile_pretrained_embeddings_spacy_en, -spacy-en) (pull_request) Failing after 7m — Build Docker (docker/Dockerfile_pretrained_embeddings_spacy_en, -spacy-en)

**Motive**
> Error processing tar file(exit status 1): write /root/.cache/pip/http/d/9/c/c/b/d9ccb85290d9889b7fd067a45d9cd98ce7bcd4ee2d66f167f5e940c2: no space left on device

degiz commented on changelog/5574.bugfix.rst:
```diff
@@ -0,0 +1 @@
+Fixed issue where metadata was not being sent with the tracker from session_start event when calling action server.
```
Fixed an issue that happened when metadata is passed in a new session.

Now the metadata is correctly passed to the ``ActionSessionStart``.

degiz commented on tests/core/test_actions.py:
```diff
@@ -510,6 +510,21 @@ async def test_action_session_start_without_slots(
     assert events == [SessionStarted(), ActionExecuted(ACTION_LISTEN_NAME)]
 
 
+async def test_action_session_start_with_metadata(
```
I don't think this test is necessary now.

lluchini replied:
Done!

degiz commented on tests/core/test_processor.py:
```diff
@@ -485,6 +485,34 @@ async def test_update_tracker_session(
+    session_event_idx = tracker.events.index(SessionStarted())
+    session_event_metadata = tracker.events[session_event_idx].metadata
+
+    assert metadata == session_event_metadata
```
assert session_event_metadata == metadata

lluchini commented:
- DeepSource is a false positive;
- Continuous Integration / Build Docker is a problem of space; and
- Continuous Integration / Run Tests can't know, it doesn't say what's wrong and I'm not able to run tests on Windows (apparently Sanic doesn't like windows very much, I'm working on it...).

lluchini commented:
> Great job! 👍
> I'm approving the PR, let's make sure the unit tests are passing in the CI.

Thanks, you made my first contribution to an open-source project so smooth, I really appreciate!

degiz commented:
@lluchini welcome to the club 🚀 

Now we're basically trying to merge the PR before the master goes ahead 😄
